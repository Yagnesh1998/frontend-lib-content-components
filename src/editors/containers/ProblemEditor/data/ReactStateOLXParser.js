import _ from 'lodash-es';
import { XMLParser, XMLBuilder } from 'fast-xml-parser';
import { ProblemTypeKeys } from '../../../data/constants/problem';

class ReactStateOLXParser {
  constructor(problemState) {
    const parserOptions = {
      ignoreAttributes: false,
      alwaysCreateTextNode: true,
    };
    const questionParserOptions = {
      ignoreAttributes: false,
      alwaysCreateTextNode: true,
      preserveOrder: true,
    };
    const questionBuilderOptions = {
      ignoreAttributes: false,
      attributeNamePrefix: '@_',
      suppressBooleanAttributes: false,
      format: true,
      preserveOrder: true,
    };
    const builderOptions = {
      ignoreAttributes: false,
      attributeNamePrefix: '@_',
      suppressBooleanAttributes: false,
      format: true,
    };
    this.questionParser = new XMLParser(questionParserOptions);
    this.parser = new XMLParser(parserOptions);
    this.builder = new XMLBuilder(builderOptions);
    this.questionBuilder = new XMLBuilder(questionBuilderOptions);
    this.editorObject = problemState.editorObject;
    this.problemState = problemState.problem;
  }

  addHints() {
    const hintsArray = [];
    const hints = _.get(this.problemState, 'settings.hints', []);
    hints.forEach(element => {
      hintsArray.push({
        '#text': element.value,
      });
    });
    const demandhint = {
      demandhint: {
        hint: hintsArray,
      },
    };
    return demandhint;
  }

  addSolution() {
    if (!_.has(this.problemState, 'settings.solutionExplanation')) { return {}; }
    const { solution } = this.editorObject;
    const parsedSolution = this.parser.parse(solution);
    const solutionObject = {
      solution: {
        ...parsedSolution,
      },
    };
    return solutionObject;
  }

  addMultiSelectAnswers(option) {
    const choice = [];
    let compoundhint = [];
    let widget = {};
    // eslint-disable-next-line prefer-const
    let { answers, generalFeedback } = this.problemState;
    const answerTitles = this.editorObject?.answers;
    // general feedback replaces selected feedback if all incorrect selected feedback is the same.
    if (generalFeedback !== ''
    && answers.every(
      answer => (
        answer.correct
          ? true
          : answer?.selectedFeedback === answers.find(a => a.correct === false).selectedFeedback
      ),
    )) {
      answers = answers.map(answer => (!answer?.correct
        ? { ...answer, selectedFeedback: generalFeedback }
        : answer));
    }
    answers.forEach((answer) => {
      const feedback = [];
      let singleAnswer = {};
      const title = answerTitles ? answerTitles[answer.id] : answer.title;
      if (this.hasAttributeWithValue(answer, 'title')) {
        if (this.hasAttributeWithValue(answer, 'selectedFeedback')) {
          feedback.push({
            '#text': _.get(answer, 'selectedFeedback'),
            '@_selected': true,
          });
        }
        if (this.hasAttributeWithValue(answer, 'unselectedFeedback')) {
          feedback.push({
            '#text': _.get(answer, 'unselectedFeedback'),
            '@_selected': false,
          });
        }
        if (this.hasAttributeWithValue(answer, 'feedback')) {
          feedback.push({
            '#text': _.get(answer, 'feedback'),
          });
        }
        if (feedback.length) {
          singleAnswer[`${option}hint`] = feedback;
        }
        singleAnswer = {
          '#text': title,
          '@_correct': answer.correct,
          ...singleAnswer,
        };
        choice.push(singleAnswer);
      }
    });
    widget = { [option]: choice };
    if (_.has(this.problemState, 'groupFeedbackList')) {
      compoundhint = this.addGroupFeedbackList();
      widget = {
        ...widget,
        compoundhint,
      };
    }
    return widget;
  }

  addGroupFeedbackList() {
    const compoundhint = [];
    const { groupFeedbackList } = this.problemState;
    groupFeedbackList.forEach((element) => {
      compoundhint.push({
        '#text': element.feedback,
        '@_value': element.answers.join(' '),
      });
    });
    return compoundhint;
  }

  addQuestion() {
    const question = this.editorObject?.question || this.problemState?.question;
    // const question = this.editorObject?.question || '';
    const questionObject = this.questionParser.parse(question);
    return questionObject;
  }

  buildMultiSelectProblem(problemType, widget, option) {
    const question = this.addQuestion();
    const widgetObject = this.addMultiSelectAnswers(option);
    const demandhint = this.addHints();
    const solution = this.addSolution();

    const problemObject = {
      problem: {
        [problemType]: {
          [widget]: widgetObject,
          ...solution,
        },
        ...demandhint,
      },
    };

    const problem = this.builder.build(problemObject);
    const questionString = this.questionBuilder.build(question);
    let problemTypeTag;
    switch (problemType) {
      case ProblemTypeKeys.MULTISELECT:
        [problemTypeTag] = problem.match(/<choiceresponse>|<choiceresponse.[^>]+>/);
        break;
      case ProblemTypeKeys.DROPDOWN:
        [problemTypeTag] = problem.match(/<optionresponse>|<optionresponse.[^>]+>/);
        break;
      case ProblemTypeKeys.SINGLESELECT:
        [problemTypeTag] = problem.match(/<multiplechoiceresponse>|<multiplechoiceresponse.[^>]+>/);
        break;
      default:
        break;
    }
    const updatedString = `${problemTypeTag}\n${questionString}`;
    const problemString = problem.replace(problemTypeTag, updatedString);

    return problemString;
  }

  buildTextInput() {
    const question = this.addQuestion();
    const demandhint = this.addHints();
    const answerObject = this.buildTextInputAnswersFeedback();
    const solution = this.addSolution();

    const problemObject = {
      problem: {
        [ProblemTypeKeys.TEXTINPUT]: {
          ...answerObject,
          ...solution,
        },
        ...demandhint,
      },
    };

    const problem = this.builder.build(problemObject);
    const questionString = this.questionBuilder.build(question);
    const [problemTypeTag] = problem.match(/<stringresponse>|<stringresponse.[^>]+>/);
    const updatedString = `${problemTypeTag}\n${questionString}`;
    const problemString = problem.replace(problemTypeTag, updatedString);

    return problemString;
  }

  buildTextInputAnswersFeedback() {
    const { answers } = this.problemState;
    const answerTitles = this.editorObject?.answers;
    let answerObject = {};
    const additionAnswers = [];
    const wrongAnswers = [];
    let firstCorrectAnswerParsed = false;
    answers.forEach((answer) => {
      const correcthint = this.getAnswerHints(answer);
      if (this.hasAttributeWithValue(answer, 'title')) {
        const title = answerTitles ? answerTitles[answer.id] : answer.title;
        if (answer.correct && firstCorrectAnswerParsed) {
          additionAnswers.push({
            '@_answer': title,
            ...correcthint,
          });
        } else if (answer.correct && !firstCorrectAnswerParsed) {
          firstCorrectAnswerParsed = true;
          answerObject = {
            '@_answer': title,
            ...correcthint,
          };
        } else if (!answer.correct) {
          wrongAnswers.push({
            '@_answer': title,
            '#text': answer.selectedFeedback,
          });
        }
      }
    });
    answerObject = {
      ...answerObject,
      additional_answer: additionAnswers,
      stringequalhint: wrongAnswers,
      '@_type': _.get(this.problemState, 'additionalAttributes.type', 'ci'),
      textline: {
        '@_size': _.get(this.problemState, 'additionalAttributes.textline.size', 20),
      },
    };
    return answerObject;
  }

  buildNumericInput() {
    const question = this.addQuestion();
    const demandhint = this.addHints();
    const answerObject = this.buildNumericalResponse();
    const solution = this.addSolution();

    const problemObject = {
      problem: {
        [ProblemTypeKeys.NUMERIC]: {
          ...answerObject,
          ...solution,
        },
        ...demandhint,
      },
    };

    const problem = this.builder.build(problemObject);
    const questionString = this.questionBuilder.build(question);
    const [problemTypeTag] = problem.match(/<numericalresponse>|<numericalresponse.[^>]+>/);
    const updatedString = `${questionString}\n${problemTypeTag}`;
    const problemString = problem.replace(problemTypeTag, updatedString);

    return problemString;
  }

  buildNumericalResponse() {
    const { answers } = this.problemState;
    const answerTitles = this.editorObject?.answers;
    let answerObject = {};
    const additionalAnswers = [];
    let firstCorrectAnswerParsed = false;
    answers.forEach((answer) => {
      const correcthint = this.getAnswerHints(answer);
      if (this.hasAttributeWithValue(answer, 'title')) {
        const title = answerTitles ? answerTitles[answer.id] : answer.title;
        if (answer.correct && !firstCorrectAnswerParsed) {
          firstCorrectAnswerParsed = true;
          let responseParam = {};
          if (_.has(answer, 'tolerance')) {
            responseParam = {
              responseparam: {
                '@_type': 'tolerance',
                '@_default': _.get(answer, 'tolerance', 0),
              },
            };
          }
          answerObject = {
            '@_answer': title,
            ...responseParam,
            ...correcthint,
          };
        } else if (answer.correct && firstCorrectAnswerParsed) {
          additionalAnswers.push({
            '@_answer': title,
            ...correcthint,
          });
        }
      }
    });
    answerObject = {
      ...answerObject,
      additional_answer: additionalAnswers,
      formulaequationinput: {
        '#text': '',
      },
    };
    return answerObject;
  }

  getAnswerHints(elementObject) {
    const feedback = elementObject?.selectedFeedback;
    let correcthint = {};
    if (feedback !== undefined && feedback !== '') {
      correcthint = {
        correcthint: {
          '#text': feedback,
        },
      };
    }
    return correcthint;
  }

  hasAttributeWithValue(obj, attr) {
    return _.has(obj, attr) && _.get(obj, attr, '').toString().trim() !== '';
  }

  buildOLX() {
    const { problemType } = this.problemState;
    let problemString = '';

    switch (problemType) {
      case ProblemTypeKeys.MULTISELECT:
        problemString = this.buildMultiSelectProblem(ProblemTypeKeys.MULTISELECT, 'checkboxgroup', 'choice');
        break;
      case ProblemTypeKeys.DROPDOWN:
        problemString = this.buildMultiSelectProblem(ProblemTypeKeys.DROPDOWN, 'optioninput', 'option');
        break;
      case ProblemTypeKeys.SINGLESELECT:
        problemString = this.buildMultiSelectProblem(ProblemTypeKeys.SINGLESELECT, 'choicegroup', 'choice');
        break;
      case ProblemTypeKeys.TEXTINPUT:
        problemString = this.buildTextInput();
        break;
      case ProblemTypeKeys.NUMERIC:
        problemString = this.buildNumericInput();
        break;
      default:
        break;
    }
    return problemString;
  }
}

export default ReactStateOLXParser;
