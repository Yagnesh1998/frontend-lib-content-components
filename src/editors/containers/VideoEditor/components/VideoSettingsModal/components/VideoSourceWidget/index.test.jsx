import React from 'react';
import { shallow } from 'enzyme';

import { formatMessage } from '../../../../../../../testUtils';
import { actions } from '../../../../../../data/redux';
import { VideoSourceWidget, mapDispatchToProps } from '.';

jest.mock('../../../../../../data/redux', () => ({
  actions: {
    video: {
      updateField: jest.fn().mockName('actions.video.updateField'),
    },
  },
  selectors: {
    video: {
      videoSource: jest.fn(state => ({ videoSource: state })),
      fallbackVideos: jest.fn(state => ({ fallbackVideos: state })),
      allowVideoDownloads: jest.fn(state => ({ allowVideoDownloads: state })),
    },
  },
}));

jest.mock('../hooks', () => ({
  selectorKeys: ['soMEkEy'],
  widgetValues: jest.fn().mockReturnValue({
    videoSource: { onChange: jest.fn(), onBlur: jest.fn(), local: '' },
    fallbackVideos: {
      formValue: ['somEUrL'],
      onChange: jest.fn(),
      onBlur: jest.fn(),
      local: '',
    },
    allowVideoDownloads: { local: false, onCheckedChange: jest.fn() },
  }),
}));

describe('VideoSourceWidget', () => {
  const props = {
    error: {},
    title: 'tiTLE',
    // inject
    intl: { formatMessage },
    // redux
    updateField: jest.fn().mockName('args.updateField'),
  };

  describe('snapshots', () => {
    test('snapshots: renders as expected with default props', () => {
      expect(
        shallow(<VideoSourceWidget {...props} />),
      ).toMatchSnapshot();
    });
  });
  describe('mapDispatchToProps', () => {
    const dispatch = jest.fn();
    test('updateField from actions.video.updateField', () => {
      expect(mapDispatchToProps.updateField).toEqual(dispatch(actions.video.updateField));
    });
  });
});
