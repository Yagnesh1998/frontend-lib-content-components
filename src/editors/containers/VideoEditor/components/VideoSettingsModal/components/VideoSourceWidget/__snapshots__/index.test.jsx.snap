// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VideoSourceWidget snapshots snapshots: renders as expected with default props 1`] = `
<injectIntl(ShimmedIntlComponent)
  title="Video source"
>
  <Form.Group>
    <div
      className="border-primary-100 border-bottom pb-4"
    >
      <Form.Control
        floatingLabel="Video ID or URL"
        onBlur={[Function]}
        onChange={[MockFunction]}
        value=""
      />
    </div>
    <div
      className="mt-3"
    >
      <FormattedMessage
        defaultMessage="Fallback videos"
        description="Title for the fallback videos section"
        id="authoring.videoeditor.videoSource.fallbackVideo.title"
      />
    </div>
    <div>
      <FormattedMessage
        defaultMessage="To be sure all learners can access the video, edX
    recommends providing additional videos in both .mp4 and
    .webm formats.  The first listed video compatible with the
    learner's device will play."
        description="Test explaining reason for fallback videos"
        id="authoring.videoeditor.videoSource.fallbackVideo.message"
      />
    </div>
    <Form.Row
      className="mt-4"
    >
      <Form.Control
        floatingLabel="Video URL"
      />
      <IconButtonWithTooltip
        alt="Delete"
        iconAs="Icon"
        key="top-delete-somEUrL"
        onClick={[Function]}
        tooltipContent="Delete"
        tooltipPlacement="top"
      />
    </Form.Row>
    <ActionRow
      className="mt-4"
    >
      <Form.Checkbox
        checked={false}
        className="decorative-control-label"
        onChange={[MockFunction]}
      >
        <div
          className="small text-gray-700"
        >
          <FormattedMessage
            defaultMessage="Allow video downloads"
            description="Label for allow video downloads checkbox"
            id="authoring.videoeditor.videoSource.allowDownloadCheckboxLabel"
          />
        </div>
      </Form.Checkbox>
      <OverLayTrigger
        key="top"
        overlay={
          <ToolTip
            id="tooltip-top"
          >
            <FormattedMessage
              defaultMessage="Allow learners to download versions of this video in
    different formats if they cannot use the edX video player or do not have
    access to YouTube."
              description="Message for allow video downloads checkbox"
              id="authoring.videoeditor.videoSource.fallbackVideo.allowDownloadTooltipMessage"
            />
          </ToolTip>
        }
        placement="top"
      >
        <Icon
          style={
            Object {
              "height": "16px",
              "width": "16px",
            }
          }
        />
      </OverLayTrigger>
      <ActionRow.Spacer />
    </ActionRow>
  </Form.Group>
  <Button
    className="text-primary-500 font-weight-bold pl-0"
    onClick={[Function]}
    size="sm"
    variant="link"
  >
    <FormattedMessage
      defaultMessage="Add a video URL"
      description="Label for add a video URL button"
      id="authoring.videoeditor.videoSource.fallbackVideo.allowDownloadTooltipMessage"
    />
  </Button>
</injectIntl(ShimmedIntlComponent)>
`;
